src/numbers_royale.cpp

```diff
@@ -0,0 +1,192 @@
+#include "numbers_royale.h"
+#include "cpu.h"
+
+
+namespace {
+  // Helper functions
+  /**
+   * @brief Wait for user to press enter
+   * 
+   * @param message Message to display
+   */
+  void wait_for_enter(const char* message = "Press <enter> to continue... ") {
+    printf("%s", message);
+    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+  }
+
+  /**
+   * @brief Get the validated input object
+   * 
+   * @param is_valid lambda function to validate input
+   * @return int selected input after validation
+   */
+  int get_validated_input(const std::function<bool(int)>& is_valid) {
```

Comment:
making this a template would be so clean though ;)
```suggestion
  template <typename T>
  T get_validated_input(const std::function<bool(T)>& is_valid) {
```

Reply:
ok, lemme learn how to make templates.

Reply:
interesting, adding templates required explicit declaration of the validation function